syntax = "proto3";

package chat;

service ChatService {
  rpc CreateAccount(CreateAccountRequest) returns (CreateAccountResponse);
  rpc Login(LoginRequest) returns (LoginResponse);
  rpc ListAccounts(ListAccountsRequest) returns (ListAccountsResponse);
  rpc SendMessage(SendMessageRequest) returns (SendMessageResponse);
  rpc ReadNewMessages(ReadNewMessagesRequest) returns (ReadNewMessagesResponse);
  rpc DeleteMessages(DeleteMessagesRequest) returns (DeleteMessagesResponse);
  rpc DeleteAccount(DeleteAccountRequest) returns (DeleteAccountResponse);
  rpc ListMessages(ListMessagesRequest) returns (ListMessagesResponse);
}

message CreateAccountRequest {
  string username = 1;
  string password = 2;
}

message CreateAccountResponse {
  string message = 1;
  bool success = 2;
}

message LoginRequest {
  string username = 1;
  string password = 2;
}

message LoginResponse {
  string message = 1;
  int32 unread_count = 2;
  bool success = 3;
}

message ListAccountsRequest {
  string username = 1;
  string pattern = 2;
}

message ListAccountsResponse {
  repeated string accounts = 1;
  bool success = 2;
}

message SendMessageRequest {
  string sender = 1;
  string to = 2;
  string content = 3;
}

message SendMessageResponse {
  string message = 1;
  bool success = 2;
}

message ReadNewMessagesRequest {
  string username = 1;
  int32 count = 2;
}

message ReadNewMessagesResponse {
  repeated string messages = 1;
  bool success = 2;
}

message DeleteMessagesRequest {
  string username = 1;
  // List of message IDs (1-indexed) to delete.
  // A value of -1 in a single-element list indicates "delete all."
  repeated int32 message_ids = 2;
}

message DeleteMessagesResponse {
  string message = 1;
  bool success = 2;
}

message DeleteAccountRequest {
  string username = 1;
}

message DeleteAccountResponse {
  string message = 1;
  bool success = 2;
}

message ListMessagesRequest {
  string username = 1;
}

message ListMessagesResponse {
  repeated string messages = 1;
  bool success = 2;
}
